Add ability for overlay to be modeless (non-modal)

limonte opened the pull request:
The overlay should have a nonmodal mode for polite components like notification.nonmodal means that a document will still remain accessible for interactions when a notification is shown.  

Part of https://github.com/vaadin/vaadin-notification/issues/4

The new minor release will be needed after this change.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/vaadin-overlay/37)
<!-- Reviewable:end -->

Commit:
Add ability for overlay to be nonmodal

tomivirkki commented:
*[vaadin-overlay.html, line 155 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxhwlMdW0J7EnbtSLZP:-KxhwlMdW0J7EnbtSLZQ:brf017q) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/07a748ea17da89e28726f13b2f711dc4b703b9cb/vaadin-overlay.html#L155)):*
> ```HTML
>           },
> 
>           modal: {
> ```

If we make modality falsy by default, won't that break every existing element that uses the overlay?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[vaadin-overlay.html, line 155 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxhwlMdW0J7EnbtSLZP:-KxiksaF1Des19Wyim73:b-rhct8p) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/07a748ea17da89e28726f13b2f711dc4b703b9cb/vaadin-overlay.html#L155)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

If we make modality falsy by default, won't that break every existing element that uses the overlay?
</blockquote></details>

No, if we release a new major (`v2.0.0`) after this change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[vaadin-overlay.html, line 155 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxhwlMdW0J7EnbtSLZP:-KxkyKkGUEnXxr_jJEJw:b-d036ia) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/07a748ea17da89e28726f13b2f711dc4b703b9cb/vaadin-overlay.html#L155)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

No, if we release a new major (`v2.0.0`) after this change.
</blockquote></details>

And after that, once we need to fix a bug in overlay, we'd have to do it for two major branches or upgrade all the other elements to use 2. Let's rather not bump major since we can easily just name  the property differently and make this an opt-in feature.

---

*[vaadin-overlay.html, line 364 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxkzMFKN5k8UqVSZ3BH:-KxkzMFKN5k8UqVSZ3BI:b4e22g6) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/07a748ea17da89e28726f13b2f711dc4b703b9cb/vaadin-overlay.html#L364)):*
> ```HTML
> 
>       _modalChanged(modal) {
>         if (modal) {
> ```

Why is the if part directly here in the function but the else part is separated to its own function while they're the same exact code (just opposite actions)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks broke.

---

*[vaadin-overlay.html, line 155 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxhwlMdW0J7EnbtSLZP:-KxlUpnJjAFLp2dX0Y1W:b-cwz8gc) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/07a748ea17da89e28726f13b2f711dc4b703b9cb/vaadin-overlay.html#L155)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

And after that, once we need to fix a bug in overlay, we'd have to do it for two major branches or upgrade all the other elements to use 2. Let's rather not bump major since we can easily just name  the property differently and make this an opt-in feature.
</blockquote></details>

Changed to `nonModal`

---

*[vaadin-overlay.html, line 364 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxkzMFKN5k8UqVSZ3BH:-KxlUtQAgVmzS3t1Iu8G:b-su6nj9) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/07a748ea17da89e28726f13b2f711dc4b703b9cb/vaadin-overlay.html#L364)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Why is the if part directly here in the function but the else part is separated to its own function while they're the same exact code (just opposite actions)?
</blockquote></details>

because we need to perform the opposite part (else) in 2 cases:

- `nonModal` has been changed
- `opened` has been changed

Both cases are covered by tests now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[vaadin-overlay.html, line 155 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxhwlMdW0J7EnbtSLZP:-KxlWcXY6jxpCNOxZx5G:b-2r2h06) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/07a748ea17da89e28726f13b2f711dc4b703b9cb/vaadin-overlay.html#L155)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

Changed to `nonModal`
</blockquote></details>

"The opposite of modal is modeless" https://en.wikipedia.org/wiki/Modal_window

---

*[vaadin-overlay.html, line 364 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxkzMFKN5k8UqVSZ3BH:-KxlXEeeDar0sm_HTypB:b-phenqf) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/07a748ea17da89e28726f13b2f711dc4b703b9cb/vaadin-overlay.html#L364)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

because we need to perform the opposite part (else) in 2 cases:

- `nonModal` has been changed
- `opened` has been changed

Both cases are covered by tests now.
</blockquote></details>

Ok, not a blocker

---

*[test/vaadin-overlay.html, line 17 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxlW9jHgqFtkCVXAGpM:-KxlW9jIGVYciIbywPZ3:b-z8jy8i) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/07f2e4889a21641d31e67ce6b8272e44aae8dae3/test/vaadin-overlay.html#L17)):*
> ```HTML
>       <div>
>         <div id="parent">
>           <vaadin-overlay modal>
> ```

This is no longer relevant

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[vaadin-overlay.html, line 155 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxhwlMdW0J7EnbtSLZP:-KxlYasemzoMP7qXSoa3:bgevqfj) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/07a748ea17da89e28726f13b2f711dc4b703b9cb/vaadin-overlay.html#L155)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

"The opposite of modal is modeless" https://en.wikipedia.org/wiki/Modal_window
</blockquote></details>

Well looks like both modeless and nonmodal are used so not a blocker.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks broke.

---

*[vaadin-overlay.html, line 155 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxhwlMdW0J7EnbtSLZP:-KxlZh1b7lVaSSPv6AAt:b-esnys8) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/07a748ea17da89e28726f13b2f711dc4b703b9cb/vaadin-overlay.html#L155)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Well looks like both modeless and nonmodal are used so not a blocker.
</blockquote></details>

Changed to `modeless` :+1:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[test/vaadin-overlay.html, line 17 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxlW9jHgqFtkCVXAGpM:-KxlZwaTVi7zSqxtpA6i:b-hfpv0l) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/07f2e4889a21641d31e67ce6b8272e44aae8dae3/test/vaadin-overlay.html#L17)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

This is no longer relevant
</blockquote></details>

removed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[vaadin-overlay.html, line 378 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxmnGNk9JADG_iUAL-o:-KxmnGNk9JADG_iUAL-p:bfigod9) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/02534ab7c0b386920823ae20222d64744708920c/vaadin-overlay.html#L378)):*
> ```HTML
>           // other document nodes
>           this._previousDocumentPointerEvents = document.body.style.pointerEvents;
>           document.body.style.pointerEvents = 'none';
> ```

This makes the whole application inaccessible (you can see this in demos). Pointer events should be disabled _only_ when the overlay is opened.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[vaadin-overlay.html, line 378 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxmnGNk9JADG_iUAL-o:-KxmvWI1pfjspoe4hQBU:b-qoli1a) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/02534ab7c0b386920823ae20222d64744708920c/vaadin-overlay.html#L378)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

This makes the whole application inaccessible (you can see this in demos). Pointer events should be disabled _only_ when the overlay is opened.
</blockquote></details>

Oops, my bad, fixed. I added `_enterModalState()` and `_exitModalState()`, both methods are called in `_openedChanged()` and `_modelessChanged()`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[vaadin-overlay.html, line 159 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxqEY6pSjcGw32iVOhr:-KxqEY6pSjcGw32iVOhs:b-8kdewp) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1a6a67031d585698ec95c1838c7144bd33fa1d06/vaadin-overlay.html#L159)):*
> ```HTML
>            * it doesn’t change the functionality of the user interface.
>            */
>           modeless: {
> ```

+1 Interesting word, I had never seen, but it actually exists. 

---

*[vaadin-overlay.html, line 402 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxqFCe38azm7pHvbTu_:-KxqFCe38azm7pHvbTua:b78uiz) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1a6a67031d585698ec95c1838c7144bd33fa1d06/vaadin-overlay.html#L402)):*
> ```HTML
>         // Restore body pointer-events
>         document.body.style.pointerEvents = this._previousDocumentPointerEvents;
>         this._previousDocumentPointerEvents = null;
> ```

Why null?  `delete  this._previousDocumentPointerEvents` seems more meaningful here. Not a blocker

---

*[demo/index.html, line 40 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxqFifeOGF464Xv7IM4:-KxqFifeOGF464Xv7IM5:b-nsgz43) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1a6a67031d585698ec95c1838c7144bd33fa1d06/demo/index.html#L40)):*
>  This overlay won't disable the main content, showing.

This sentence sounds incorrect. I think you can remove the `,showing` part. Seems a mistake when copying pasting from above

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[vaadin-overlay.html, line 159 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxqEY6pSjcGw32iVOhr:-KxqTPUwO7JdP_PRjx6a:be64brv) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1a6a67031d585698ec95c1838c7144bd33fa1d06/vaadin-overlay.html#L159)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

+1 Interesting word, I had never seen, but it actually exists. 
</blockquote></details>

Tomi found it in Wikipedia :D

---

*[vaadin-overlay.html, line 402 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxqFCe38azm7pHvbTu_:-KxqTUA7EL_72r9AVtcw:b-896fix) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1a6a67031d585698ec95c1838c7144bd33fa1d06/vaadin-overlay.html#L402)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Why null?  `delete  this._previousDocumentPointerEvents` seems more meaningful here. Not a blocker
</blockquote></details>

Done.

---

*[demo/index.html, line 40 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxqFifeOGF464Xv7IM4:-KxqTWNeDNGM0xEwC5JG:b-896fix) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1a6a67031d585698ec95c1838c7144bd33fa1d06/demo/index.html#L40)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

>  This overlay won't disable the main content, showing.

This sentence sounds incorrect. I think you can remove the `,showing` part. Seems a mistake when copying pasting from above
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[vaadin-overlay.html, line 159 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-KxqEY6pSjcGw32iVOhr:-KxqU1Zfv67XXL-3MYEi:b-3mnw1x) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1a6a67031d585698ec95c1838c7144bd33fa1d06/vaadin-overlay.html#L159)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

Tomi found it in Wikipedia :D
</blockquote></details>

Learn how to Google:
 - dialog nonmodal: 82500 results
 - dialog modeless: 135000 results

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37)*
<!-- Sent from Reviewable.io -->

manolo commented:
Reviewed 1 of 3 files at r3, 1 of 3 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37)*
<!-- Sent from Reviewable.io -->

manolo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-:-KxrqzSMOtZTXFm2Uvah:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r3, 1 of 3 files at r4, 2 of 2 files at r5.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-overlay/37#-:-KxruvvPlpxtYU3Lb8Ne:bnfp4nl)*
<!-- Sent from Reviewable.io -->
